src/cpu/testers/spatter_gen/spatter_gen.cc

```diff
@@ -0,0 +1,584 @@
+
+            // occupy on fp register
+            available_fp_regs_this_cycle--;
+            fp_reg_used_delta++;
+            // make AGU busy for the next requestGenLatency cycles.
+            generatorBusyUntil[i] = clockEdge(Cycles(requestGenLatency));
+
+            size_t value_size;
+            Addr value_addr;
+            std::tie(value_size, value_addr) = idx_info->getAccessPair();
+            // since this is the last part of access _index_value does not
+            // matter.
+            AccessInfo* acc_info = new AccessInfo(
+                                                AccessType::VALUE,
+                                                idx_info->kernel_type(),
+                                                -1,
+                                                -1
+                                                );
+            acc_info->trip_time(idx_info->trip_time());
+            // if from a scatter kernel do write, otherwise do read.
+            // TODO: this should be made into a function in SpatterKernel
+            // or AccessInfo.
+            MemCmd cmd =
+                    idx_info->kernel_type() == SpatterKernelType::scatter ?
+                    MemCmd::WriteReq : MemCmd::ReadReq;
+
+            // create the request for the address and size
+            RequestPtr req = std::make_shared<Request>(
+                                                    value_addr,
+                                                    value_size,
+                                                    0,
+                                                    requestorId
+                                                    );
+            // Dummy PC to have PC-based prefetchers latch on;
+            // get entropy into higher bits
+            req->setPC(((Addr)requestorId) << 2);
+
+            // finally create the packet and allocate host memory for data.
+            PacketPtr pkt = new Packet(req, cmd);
+            uint8_t* pkt_data = new uint8_t[req->getSize()];
+            // Randomly intialize pkt_data, for testing cache coherence.
+            for (int i = 0; i < req->getSize(); i++) {
+                pkt_data[i] = random_mt.random<uint8_t>();
+            }
+            pkt->dataDynamic(pkt_data);
+            pkt->pushSenderState(acc_info);
+
+            // push to requestBuffer
+            requestBuffer.push(pkt);
+            requestBufferInsertionTime.push(curTick());
+            DPRINTF(
+                SpatterGen,
+                "%s: Pushed pkt: %s to requestBuffer. Access to VALUES.\n",
+                __func__, pkt->print()
+            );
+            // now deallocate resources for reading the index
+            delete idx_info;
+            delete idx_pkt;
+            int_reg_used_delta--;
+            receiveBuffer.pop();
+            receiveBufferInsertionTime.pop();
+        } else if (have_kernel && have_int_reg) {
```

Comment:
I would split these two `if`/`else if` clauses into their own functions. This function is too big.

Reply:
I agree. I would argue for not changing it right now. I will update this after we have a more definitive picture of what we want the SpatterKernel class to look like.

Reply:
I would argue that if we don't know what we want the kernel to look like, then it's not ready to be merged upstream. We should go ahead and do it the right way the first time. Once something is merged, the activation energy to change it is very high.